faker/providers/barcode/__init__.py

```diff
@@ -23,8 +58,157 @@ def ean(self, length=13):
+        elif base[-1] <= 4:
+            code = base[:base[-1]] + [0] * 5 + base[base[-1]:-1]
+        else:
+            code = base[:5] + [0] * 4 + base[-1:]
+
+        code.insert(0, number_system_digit)
+        weights = [3, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3]
+        weighted_sum = sum(x * y for x, y in zip(code, weights))
+        check_digit = (10 - weighted_sum % 10) % 10
+        code.append(check_digit)
+        return ''.join(str(x) for x in code)
+
+    def ean(self, length=13):
+        return self._ean(length)
+
     def ean8(self):
-        return self.ean(8)
+        return self._ean(8)
+
+    def ean13(self, *, no_leading_zero=False, force_leading_zero=False):
+        """
+        Creates an EAN-13 barcode
+
+        Notes on this method:
+        - Obviously, `no_leading_zero` and `force_leading_zero` cannot be both set to True.
+        - If `no_leading_zero` is True, no EAN-13 barcodes generated will lead with zero. This flag
+          can be used to work around scanners that cannot properly handle such barcodes.
+        - If `force_leading_zero` is True, EAN-13 barcodes generated will always lead with zero,
+          and as a result, the barcodes become UPC-A compatible. Just drop the leading zeroes.
+        - Any other non-boolean value supplied to `no_leading_zero` and `force_leading_zero` will
+          be treated as False, and therefore, providing the default behavior.
+
+        :param no_leading_zero: True or False
+        :param force_leading_zero: True or False
+        :return: An EAN-13 barcode
+        """
+        return self._ean(13, no_leading_zero=no_leading_zero, force_leading_zero=force_leading_zero)
+
+    def upc_a(self, *, upc_ae_mode=False, base=None, number_system_digit=None):
+        """
+        Creates a 12-digit UPC-A barcode
+
+        Notes on UPC-A barcode and this method:
+        - EAN-13 is a superset of UPC-A. Simply put, leading zero + UPC-A = EAN-13.
+        - For the lack of a concise and better term, this provider uses the term "upc_ae" to mean
+          that a UPC-A barcode can be converted to UPC-E.
+        - If `upc_ae_mode` is enabled, this method will only return UPC-A barcodes that can be
+          converted to UPC-E. In this mode the leading digit (number system) of the UPC-A barcode
+          may only start with 0 or 1.
+        - When `upc_ae_mode` is enabled, the `number_system_digit` may be explicitly set to 0 or 1.
+          If it is not set or is invalid, either values will be chosen at random.
+        - When `upc_ae_mode` is enabled, a 6-digit UPC-E string `base` may be explicitly set. If it
+          is not set or is invalid, a random 6 digit combination will be used.
+        - If `upc_ae_mode` is disabled, the values of `base` and `number_system_digit` are ignored.
+
+        :param upc_ae_mode: Set to True explicitly to enable
+        :param base: A 6-digit string
+        :param number_system_digit: 0, 1, '0', or '1'
```

Comment:
I'd rather avoid casting down the line and just make sure this argument is a number to begin with. We can add an `assert isinstance(number_system_digit, int)` as a guard on input

Reply:
Alright, I will change that to an integer.